extensions/github/src/components/WorkflowForm.tsx

```diff
@@ -0,0 +1,269 @@
+    });
+    return [];
+  }
+
+  if (!workflowDispatch.inputs) {
+    return [];
+  }
+
+  const inputs = workflowDispatch.inputs;
+  const inputsList: Input[] = [];
+  for (const key in inputs) {
+    if (Object.prototype.hasOwnProperty.call(inputs, key)) {
+      const value = inputs[key];
+      inputsList.push({
+        name: key,
+        description: value.description,
+        default: value.default,
+        required: value.required,
+        type: value.type,
+        options: value.options,
+      });
+    }
+  }
+  return inputsList;
+}
+
+/**
+ * Validates if all required workflow inputs are present in the form values.
+ *
+ * @param inputs The inputs in the workflow.
+ * @param values The values in the form.
+ * @returns Returns true if all required inputs are present in the form's values object. If any required input is missing, returns false.
+ */
+function validateFormValues(inputs: Input[], values: Form.Values) {
+  const requiredInputs = inputs.filter((input) => input.required);
+  for (const input of requiredInputs) {
+    if (!values[input.name]) {
+      throw new Error(`Input ${input.name} is required`);
+    }
+  }
+  return true;
+}
+
+export function WorkflowForm({ repository, workflow, branches }: WorkflowFormProps) {
+  const { github } = getGitHubClient();
+  const { octokit } = getGitHubClient();
+
+  const [owner, repo] = repository.split("/");
+  const { data: repositories } = useMyRepositories();
+  const { data: repositoryData, isLoading: repositoryIsLoading } = useCachedPromise(
+    (repository) => {
+      const selectedRepository = repositories?.find((r) => r.name === repo);
+
+      if (!selectedRepository) {
+        return Promise.resolve(null);
+      }
+      return github.dataForRepository({ owner: selectedRepository.owner.login, name: selectedRepository.name });
+    },
+    [repository]
+  );
+
+  const [selectedBranch, setSelectedBranch] = useState<string>(
+    repositoryData?.repository?.defaultBranchRef?.name ?? ""
+  );
+
+  const { data: inputData, isLoading: inputIsLoading } = useCachedPromise(
+    async (workflow) => {
+      try {
+        const content = await getWorkflowContent(octokit, repo, owner, workflow);
+        return await getWorkflowInputs(content);
+      } catch {
+        return [];
+      }
+    },
+    [workflow]
+  );
+
+  /**
+   * Stores the input error strings.
+   */
+  const [inputErrors, setInputErrors] = useState<{ [name: string]: string }>({});
+
+  /*
+   * Updates the error messages based on the form value for the specified input.
+   */
+  const updateInputErrors = (name: string, value: string | undefined) => {
+    const input = inputData?.find((input) => input.name === name);
+    if (input && input.required && !value) {
+      setInputErrors((prevState) => ({ ...prevState, [name]: `${name} is required` }));
+    } else {
+      setInputErrors((prevState) => ({ ...prevState, [name]: "" }));
+    }
+  };
+
+  async function onSubmit(values: Form.Values) {
```

Comment:
You can probably use the `useForm` hook from the utils package. This makes error management easier to handle 😊

Reply:
Ah I didn't know about this one! I took a look at that, and I'm not sure exactly how to use it if the form values are unknown when I first create the firm. It looks like there is a `validation` dictionary of fields that you can pass in and validate. Since I am getting the Form values by parsing the yml file inputs, I don't actually know those values at initialization time. 

Is there a way to still use the `useForm` hook, without knowing the values to validate up front?

Reply:
Sure, it's possible! What you could do is set up an object containing validation rules that is initially empty using `useState` and pass it to `useForm`. Then you can parse the yml files and replace the initial empty object with the validation rules. You can find an example in `extensions/jira/src/components/CreateIssueForm.tsx` that does just that:

```tsx
export default function CreateIssueForm(props: CreateIssueFormProps) {
  const [customFieldInitialValues, setCustomFieldInitialValues] = useState({});
  const [customFieldsValidation, setCustomFieldsValidation] = useState({});

  const { handleSubmit, itemProps, values, setValue, reset, focus } = useForm<IssueFormValues>({
    async onSubmit(values) {
      // ...
    },
    initialValues:  {
      // ...
    },
    validation: {
      projectId: FormValidation.Required,
      issueTypeId: FormValidation.Required,
      summary: FormValidation.Required,
      ...customFieldsValidation,
    },
  });

  const { unknownCustomFields, customFields } = useMemo(() => {
    if (selectedIssueType) {
      const { unknownCustomFields, customFields, initialValues, validation } =
        getCustomFieldsForCreateIssue(selectedIssueType);

      setCustomFieldInitialValues(initialValues);
      setCustomFieldsValidation(validation);

      return { unknownCustomFields, customFields };
    }

    return { unknownCustomFields: "", customFields: [] };
  }, [selectedIssueType]);


  return (
    // ...
  )
}
```

Let me know if you need extra help with that.

Reply:
@thomaslombart Ok, I kinda got this working with this code:

```
  const [customFieldInitialValues, setCustomFieldInitialValues] = useState({});
  const [customFieldsValidation, setCustomFieldsValidation] = useState({});

  const { handleSubmit } = useForm<Form.Values>({
    async onSubmit(values) {
      try {
        if (!inputData) {
          return;
        }
  
        // Remove the branch from the form values, and pass the remaining values as inputs.
        delete values.branch;
  
        validateFormValues(inputData, values);
  
        await runWorkflow(workflow, repository, selectedBranch, values);
      } catch (error) {
        await showToast({
          style: Toast.Style.Failure,
          title: "Inputs are invalid",
          message: getErrorMessage(error),
        });
      }
    },
    initialValues:  {
      ...customFieldInitialValues
    },
    validation: {
      ...customFieldsValidation,
    },
  });

  const { data: inputData, isLoading: inputIsLoading } = useCachedPromise(
    async (workflow) => {
      const content = await getWorkflowContent(octokit, repo, owner, workflow);
      return await getWorkflowInputs(content);
    },
    [workflow]
  );

  // Set the initial values of the form to the yml file inputs.
  useEffect(() => {
    if (inputData) {
      const initialValues = inputData.reduce<Record<string, string | boolean>>((acc, input) => {
        acc[input.name] = input.default;
        return acc;
      }, {});
      setCustomFieldInitialValues(initialValues);
    }
  }, [inputData]);

  type FormValidationRule = {
    required?: boolean;
  };
  
  type FormValidationRules = Record<string, FormValidationRule>;

  // Set the initial validation rules based on the required values of the yml file inputs.
  useEffect(() => {
    if (inputData) {
      const validations = inputData.reduce<FormValidationRules>((acc, input) => {
        if (input.required) {
          acc[input.name] = { required: true };
        }
        return acc;
      }, {});
      setCustomFieldsValidation(validations);
    }
  }, [inputData]);
```

The submit validation is working fine, but the problem is when I delete the text in one of the required fields, the error message doesn't show up immediately.

Before, I was using onBlur to handle the error message updates:

```
                <Form.TextField
                  key={input.name}
                  id={input.name}
                  title={input.name}
                  defaultValue={input.default}
                  // onBlur={(event) => {
                  //   updateInputErrors(input.name, event.target.value);
                  // }}
                  error={inputErrors[input.name]}
                />
```

Now, I think useForm is supposed to automatically update the message. I must be missing something in the validation set up. I don't see in `create-issue.tsx` where the extra validations are set. Am I looking in the right place?

Thanks!

Reply:
Could you share a recording so that I can see what's happening? Can you also update your branch so that I can test the code locally? 

Note that you shouldn't use `validateFormValues(inputData, values);` in `onSubmit` as this callback won't be run unless all values are properly validated. Also, you shouldn't have to pass `onBlur` before. `useForm` is returning `itemProps` for you which you can use in your different Form components like the following example:

```tsx
export type IssueFormValues = {
  projectId: string;
  // ...
};

const { handleSubmit, itemProps, ... } = useForm<IssueFormValues>({ ... })

<Form
  actions={
    <ActionPanel>
      <Action.SubmitForm onSubmit={handleSubmit} title="Create Issue" icon={Icon.Plus} />
    </ActionPanel>
  }
>
  <Form.Dropdown {...itemProps.projectId} title="Project" storeValue>
    {projects?.map((project) => {
      return (
        <Form.Dropdown.Item
          key={project.id}
          value={project.id}
          title={`${project.name} (${project.key})`}
          icon={project.avatarUrls["32x32"]}
        />
      );
    })}
  </Form.Dropdown>
  // ...
</Form>
```

Reply:
Pushed the changes to this branch so you can test. Here's the commit: https://github.com/raycast/extensions/pull/5214/commits/517055b0b55cffa5ff6639cf8e366bb32b63e51f

Here's a video of the validation not running when the field focus changes:

https://user-images.githubusercontent.com/756119/229023782-e117fd46-2442-4413-92d2-9b6dcfbada93.mp4

Before I updated the code, when you deleted the input in `name`, an error would appear that says "Please enter a name". Now, I must be missing something because the validation is not running using the `customFieldsValidation` I added. 

Also, I removed the `validateFormValues()` call, and now I can submit the form even if some of the required fields are empty. Before I removed that, `validateFormValues()` was catching those errors. Probably the same issue where `customFieldsValidation` is not set properly. Or, I'm setting in the wrong way?

Reply:
Also, in the last commit, `interface WorkflowFormValues {}` is empty, but should I add the selected branch as part of the form values? Currently, it's passed to `runWorkflow()` in `onSubmit`, but it's not validated with the rest of the form values. 

It's a required field, too. If the `branches` that are passed to `WorkflowForm` on initialization is empty, then I don't think we can send a run workflow command.

Reply:
Hey @josh-berlin, `useForm` returns an `itemProps` object that you need to pass to your different dropdown items. It doesn't seem to be there as of now. This `itemProps` object contains a bunch of things like the form ID, the value, the error field, etc. Here's an example from the developers docs that should lead you to the right code: https://developers.raycast.com/utilities/react-hooks/useform#example 😊